projects/l2_pooling/multi_column_convergence.py

```diff
@@ -38,54 +40,91 @@
   createObjectMachine
 )
 
-def locateConvergencePoint(stats, targetValue):
+def locateConvergencePoint(stats, minOverlap, maxOverlap):
   """
   Walk backwards through stats until you locate the first point that diverges
-  from targetValue.  We need this to handle cases where it might get to
-  targetValue, diverge, and then get back again.  We want the last convergence
-  point.
+  from target overlap values.  We need this to handle cases where it might get
+  to target values, diverge, and then get back again.  We want the last
+  convergence point.
   """
   for i,v in enumerate(stats[::-1]):
-    if v != targetValue:
-      return len(stats)-i
+    if not (v >= minOverlap and v <= maxOverlap):
+      return len(stats)-i + 1
 
-  # Never differs - converged right away
-  return 0
+  # Never differs - converged in one iteration
+  return 1
 
 
-def averageConvergencePoint(inferenceStats, prefix, targetValue):
+def averageConvergencePoint(inferenceStats, prefix, minOverlap, maxOverlap,
+                            settlingTime):
   """
+  inferenceStats contains activity traces while the system visits each object.
+
+  Given the i'th object, inferenceStats[i] contains activity statistics for
+  each column for each region for the entire sequence of sensations.
+
+  For each object, compute the convergence time - the first point when all
+  L2 columns have converged.
+
+  Return the average convergence time across all objects.
+
   Given inference statistics for a bunch of runs, locate all traces with the
   given prefix. For each trace locate the iteration where it finally settles
   on targetValue. Return the average settling iteration across all runs.
   """
-  itSum = 0
-  itNum = 0
+  convergenceSum = 0.0
+
+  # For each object
   for stats in inferenceStats:
+
+    # For each L2 column locate convergence time
+    convergencePoint = 0.0
     for key in stats.iterkeys():
       if prefix in key:
-        itSum += locateConvergencePoint(stats[key], targetValue)
-        itNum += 1
+        convergencePoint = max(
+          convergencePoint,
+          locateConvergencePoint(stats[key], minOverlap, maxOverlap))
```

Comment:
Hmm, this might be a bug. In these cases the L2 columns are supposed to converge. In a clean case like this, once a column has inferred one object unambiguously, it should never increase back up to two.